### ISSUE TITLE
Nan losses with Gemma 1 DPO training on Kaggle TPU

### ISSUE THREAD

**Issue by defdet** [erfanzar/EasyDeL#170]
Nan losses when training:
![image](https://github.com/user-attachments/assets/78126797-27e6-433c-91bb-cf8260302e6c)

Please take a look at this code:

```
!pip install jax[tpu]==0.4.28 -f https://storage.googleapis.com/jax-releases/libtpu_releases.html -qq -U
!pip install datasets gradio wandb sentencepiece -U -qq
!apt-get update -qq && apt-get upgrade -y -qq && apt-get install golang -y -qq
!pip install git+https://github.com/huggingface/transformers -qq
!pip install easydel==0.0.67 -qq

from easydel import (
    AutoEasyDeLModelForCausalLM,
    EasyDeLGradientCheckPointers,
    EasyDeLOptimizers,
    EasyDeLSchedulers,
    EasyDeLState,
    SFTTrainer,
    TrainArguments,
    conversations_formatting_function,
    easystate_to_huggingface_model,
    get_modules_by_type,
)
from easydel import DPOTrainer, PartitionAxis

from transformers import AutoTokenizer
from datasets import load_dataset

fn = lambda x:[conversations_formatting_function(tokenizer, messages_field="messages")(x)]
tokenizer_name = "Defetya/gemma-2b-ru"

def gemma_format(example):
    prompt = tokenizer.apply_chat_template(example['prompt'], tokenize=False, add_generational_prompt=True)
    chosen = tokenizer.apply_chat_template(example['chosen'], tokenize=False, add_generational_prompt=True,)
    rejected = tokenizer.apply_chat_template(example['rejected'], tokenize=False, add_generational_prompt=True)
    return {
        "prompt": prompt,
        "chosen": chosen,
        "rejected": rejected,
    }
tokenizer = AutoTokenizer.from_pretrained(
    tokenizer_name,
    trust_remote_code=True
)
dataset = load_dataset('IlyaGusev/saiga_preferences')
dataset = dataset['train'].map(gemma_format, remove_columns=['prompt', 'chosen', 'rejected', 'chosen_model', 'rejected_model', 'source', 'sonnet_approved']).select(range(1_000))

import jax.numpy as jnp

model_name = "Defetya/gemma-2b-ru"
ref_model_name = model_name

max_length = 4056  # Overall maximum length
max_target_length = 1024  # Maximum Length for target column in Dataset
max_prompt_length = 512  # Maximum Length for prompt column in Dataset

dtype = jnp.bfloat16

sharding_axis_dims = (1, -1, 1, 1)
sharding_axis_names = ("dp", "fsdp", "tp", "sp")


tokenizer.pad_token = tokenizer.eos_token
tokenizer.padding_side = "left"

train_args = TrainArguments(
    model_name="EasyDeL-DPO",
    num_train_epochs=1,
    learning_rate=1e-7,
    learning_rate_end=8e-8,
    warmup_steps=0,
    optimizer=EasyDeLOptimizers.ADAMW,
    scheduler=EasyDeLSchedulers.LINEAR,
    weight_decay=0.02,
    total_batch_size=8,
    max_sequence_length=max_length,
    gradient_checkpointing=EasyDeLGradientCheckPointers.NOTHING_SAVEABLE,
    sharding_array=(1, -1, 1, 1),
    gradient_accumulation_steps=1,
    dtype=dtype,
    backend='tpu',
    param_dtype=dtype,
    training_time="7H",
    do_train=True,
    do_eval=False,
    track_memory=True
)

tokenizer.pad_token = tokenizer.eos_token if tokenizer.pad_token is None else tokenizer.pad_token
tokenizer.pad_token_id = tokenizer.eos_token_id if tokenizer.pad_token_id is None else tokenizer.pad_token_id

model_state = EasyDeLState.from_pretrained(
    pretrained_model_name_or_path=ref_model_name,
    dtype=dtype,
    param_dtype=dtype,
    init_optimizer_state=False,
    free_optimizer_state=True,
    sharding_axis_dims=sharding_axis_dims,
    sharding_axis_names=sharding_axis_names,
    partition_axis=PartitionAxis(
        batch_axis=("dp", "fsdp"),
        query_sequence_axis="sp",
        key_sequence_axis="sp",
        head_axis="tp",
        attention_dim_axis=None
    ),
    config_kwargs=dict(
        gradient_checkpointing="nothing_saveable",
        use_scan_mlp=False,
        attn_mechanism = "sharded_vanilla"
    )
)

print("Model State is Loaded.")

ref_model_state = EasyDeLState.from_pretrained(
    pretrained_model_name_or_path=model_name,
    dtype=dtype,
    param_dtype=dtype,
    init_optimizer_state=False,
    free_optimizer_state=True,
    sharding_axis_dims=sharding_axis_dims,
    sharding_axis_names=sharding_axis_names,
    partition_axis=PartitionAxis(
        batch_axis=("dp", "fsdp"),
        query_sequence_axis="sp",
        key_sequence_axis="sp",
        head_axis="tp",
        attention_dim_axis=None
    ),
    config_kwargs=dict(
        gradient_checkpointing="nothing_saveable",
        use_scan_mlp=False,
        attn_mechanism = "sharded_vanilla"

    ),
    load_in_8bit=False
)
print("Ref Model State is Loaded.")

dpo_trainer = DPOTrainer(
    model_state=model_state,
    ref_model_state=ref_model_state,
    arguments=train_args,
    train_dataset=dataset,
    tokenizer=tokenizer,
    beta=0.1,
    max_target_length=max_target_length,
    max_prompt_length=max_prompt_length,
    max_length=max_length,
)

dpo_trainer.train()

```

I've tried to lower LR, doesn't help. Any assistance would be appreciated.

**Comment by defdet** [comment]
It's basically this [code](https://github.com/erfanzar/EasyDeL/blob/main/examples/training/dpo_training_example.py), with addition of sharded_vanilla attention algorithm and adaptation for my data

**Comment by erfanzar** [comment]
hi, sorry for being late.
i guess this might be the model weight itself issue, ill try and update that if I find any issues

**Comment by defdet** [comment]
Thank you

**Comment by erfanzar** [comment]
@defdet  is the issue fixed??

**Comment by defdet** [comment]
@erfanzar I can confirm that Nan losses issue is indeed fixed with the new example. However, there is a different issue: with "ipo" loss type, loss value becomes way too big (in tens of thousands range) and perplexity becomes inf. Switching to sigmoid seems to fix the issue however.
